mautic/mautic
# Handle salesforce deletes gracefully

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? |
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

If a SF field is mapped in Mautic then later deleted, Mautic will continue to use that field in SF queries leading to failure to sync due to SF returning an error the field does not exist.

```
INTEGRATION ERROR: Salesforce - Job_Level__c, Lead_Company__c, example_Score__c__Contact, Prospecting_Qualification_Status__c ^ ERROR at Row:1:Column:113 No such column 'example_Score__c__Contact' on entity 'Lead'. If you are attempting to use a custom field, be sure to append the '__c' after the custom field name. Please reference your WSDL or the describe call for the appropriate names.
```

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. Setup Salesforce integration (https://www.mautic.org/docs/en/plugins/salesforce.html)
2. Create a text field in Salesforce for testing purposes, but it doesn't matter which field type you use. Or can use an existing field (not the obligatory one like `email`).
3. Go to Salesforce plugin settings, map your Salesforce field (contact or lead) to some Mautic field.
4. Save settings.
5. Go to your Salesforce account and delete the field you've just mapped.
6. Run `php app/console mautic:integration:fetchleads --integration=Salesforce --time-interval="10 days"`.
7. You will get an error in the console about missing column (see description above).

#### Steps to test this PR:
1. Load up this PR
2.  Repeat steps 1-6 from "Steps to reproduce the bug".
3. You should get no error in the terminal.
4. Go to Salesforce plugin settings and verify that the field you've just deleted is no longer mapped.
